### PR TITLE
NN-1415 : Surface User details

### PR DESCRIPTION
**Pull request by Mjwillis**

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/nomis/api/Caseload.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.justice.digital.nomis.api;
+
+import lombok.Builder;
+import lombok.Data;
+
+import java.util.List;
+
+@Data
+@Builder
+public class Caseload {
```
Comment (llmikeyj) [inline src/main/java/uk/gov/justice/digital/nomis/api/Caseload.java]: Can you explain what a caseload is? It looks like a named list of agency locations?

File: src/main/java/uk/gov/justice/digital/nomis/api/UserDetail.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package uk.gov.justice.digital.nomis.api;
+
+import lombok.Builder;
+import lombok.Data;
+
+import java.util.List;
+
+@Data
+@Builder
+public class UserDetail {
```
Comment (llmikeyj) [inline src/main/java/uk/gov/justice/digital/nomis/api/UserDetail.java]: Can you explain what a UserDetail is? It looks like a collection of caseloads?

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyIdentity.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+@EqualsAndHashCode(of = {"agencyLocationId", "caseload"})
+@Embeddable
+@ToString(of = {"agencyLocationId", "caseload"})
+public class CaseloadAgencyIdentity implements Serializable {
+
+    @Column(name = "AGY_LOC_ID")
+    private String agencyLocationId;
+
+    @Column(name = "caseload_id")
+    private String caseload;
```
Comment (llmikeyj) [inline src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyIdentity.java]: caseloadId, Shurely?

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyIdentity.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import java.io.Serializable;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder(toBuilder = true)
+@EqualsAndHashCode(of = {"agencyLocationId", "caseload"})
+@Embeddable
+@ToString(of = {"agencyLocationId", "caseload"})
+public class CaseloadAgencyIdentity implements Serializable {
```
Comment (llmikeyj) [inline src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyIdentity.java]: Isn't this class just modelling a join table? Do we need it, or can we do clevers with the JPA?

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package uk.gov.justice.digital.nomis.jpa.entity;
+
+import lombok.*;
+
+import javax.persistence.*;
+import java.io.Serializable;
+
+@Entity
+@Table(name = "CASELOAD_AGENCY_LOCATIONS")
+@Data()
```
Comment (llmikeyj) [inline src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java]: ()

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package uk.gov.justice.digital.nomis.jpa.entity;
+
+import lombok.*;
```
Comment (llmikeyj) [inline src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java]: Not a fan of wildcards :/

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+@AllArgsConstructor
+@Builder(toBuilder = true)
+@EqualsAndHashCode(of = {"caseload", "agencyLocation"})
+@ToString(of = {"caseload", "agencyLocation", "updateAllowedFlag"})
+public class CaseloadAgencyLocation implements Serializable {
+
+    @EmbeddedId
+    private CaseloadAgencyIdentity id;
+
+    @OneToOne(fetch = FetchType.LAZY)
```
Comment (llmikeyj) [inline src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java]: Why explicitly lazy on a one to one?

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    @JoinColumn(name = "CASELOAD_ID", updatable = false, insertable = false)
```
Comment (llmikeyj) [inline src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java]: I don't think updatable and insertable are necessary where we are not also mapping the column?

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/StaffUserAccount.java [hunk 9]
```diff
@@ -1,22 +1,8 @@
 package uk.gov.justice.digital.nomis.jpa.entity;
 
-import lombok.AllArgsConstructor;
-import lombok.Builder;
-import lombok.Data;
-import lombok.EqualsAndHashCode;
-import lombok.NoArgsConstructor;
-import lombok.ToString;
+import lombok.*;
```
Comment (llmikeyj) [inline src/main/java/uk/gov/justice/digital/nomis/jpa/entity/StaffUserAccount.java]: As previous

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/StaffUserAccount.java [hunk 10]
```diff
@@ -1,22 +1,8 @@
...
-import javax.persistence.Column;
-import javax.persistence.Entity;
-import javax.persistence.Id;
-import javax.persistence.JoinColumn;
-import javax.persistence.ManyToOne;
-import javax.persistence.OneToMany;
-import javax.persistence.OneToOne;
-import javax.persistence.PrimaryKeyJoinColumn;
-import javax.persistence.Table;
+import javax.persistence.*;
```
Comment (llmikeyj) [inline src/main/java/uk/gov/justice/digital/nomis/jpa/entity/StaffUserAccount.java]: As lombok

File: src/main/java/uk/gov/justice/digital/nomis/service/UsersService.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import uk.gov.justice.digital.nomis.jpa.repository.StaffUserAccountRepository;
+import uk.gov.justice.digital.nomis.service.transformer.UserTransformer;
+
+import java.util.Optional;
+
+@Service
+public class UsersService {
+
+    private final UserTransformer userTransformer;
+    private final StaffUserAccountRepository staffUserAccountRepository;
```
Comment (llmikeyj) [inline src/main/java/uk/gov/justice/digital/nomis/service/UsersService.java]: Where is this StaffUserAccountRepository? Is it in this review?

File: src/main/resources/R__3_3__CASELOAD_AGENCY_LOCATIONS.sql [hunk 12]
```diff
@@ -0,0 +1,27 @@
+INSERT INTO CASELOAD_AGENCY_LOCATIONS (CASELOAD_ID, AGY_LOC_ID)
+  VALUES ('BXI', 'BXI');
```
Comment (llmikeyj) [inline src/main/resources/R__3_3__CASELOAD_AGENCY_LOCATIONS.sql]: I'm confused. Caseload Id is clearly an agency location id?

Comment (Mjwillis) [reply src/main/resources/R__3_3__CASELOAD_AGENCY_LOCATIONS.sql]: No it’s not I’m afraid- it’s a one to many relationship - but will discuss with you tomorrow

Comment (Mjwillis) [reply src/main/resources/R__3_3__CASELOAD_AGENCY_LOCATIONS.sql]: No it’s not I’m afraid- it’s a one to many relationship - but will discuss with you tomorrow

File: src/main/java/uk/gov/justice/digital/nomis/api/Caseload.java (hunk 1 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/nomis/api/Caseload.java]: A caseload a collection of agencies and determines what a user can "see" in the system. Its normally a one-to-one relationship except in the Isle of Whight and central caseload

File: src/main/java/uk/gov/justice/digital/nomis/api/UserDetail.java (hunk 2 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/nomis/api/UserDetail.java]: UserDetail renamed to StaffUser - Its a member of staff, I've now added their name and currrent active status

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyIdentity.java (hunk 3 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyIdentity.java]: renamed

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyIdentity.java (hunk 4 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyIdentity.java]: No, it has extra attributes

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java (hunk 5 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java]: Again their are extra attributes

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java (hunk 6 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java]: But all our devs use the standard IntelliJ Ultimate optimiser upon checkin - > 5 statics are grouped as a wildcard.  Don't want to have a "special" set of rules just for this project.  Consistency is more important.

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java (hunk 7 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java]: I like to make every lazy - prevents joins on big query selects

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java (hunk 8 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyLocation.java]: We are mapping the column in the ID

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/StaffUserAccount.java (hunk 9 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/nomis/jpa/entity/StaffUserAccount.java]: See above

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/StaffUserAccount.java (hunk 10 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/nomis/jpa/entity/StaffUserAccount.java]: See above

File: src/main/java/uk/gov/justice/digital/nomis/service/UsersService.java (hunk 11 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/digital/nomis/service/UsersService.java]: It was in a previous review you did

File: src/main/resources/R__3_3__CASELOAD_AGENCY_LOCATIONS.sql (hunk 12 above)
Comment (Mjwillis) [reply src/main/resources/R__3_3__CASELOAD_AGENCY_LOCATIONS.sql]: I've added some examples of 1 to many

**Comment by Mjwillis** [comment]
@llmikeyj can you review and see comments.

File: src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyIdentity.java (hunk 4 above)
Comment (llmikeyj) [reply src/main/java/uk/gov/justice/digital/nomis/jpa/entity/CaseloadAgencyIdentity.java]: ok

**Pull request merged**